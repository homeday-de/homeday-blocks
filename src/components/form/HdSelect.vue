<template>
  <div
    :class="fieldClasses"
    class="field field--select">
    <img
      v-if="icon"
      :src="icon"
      role="presentation"
      class="field__icon"
    >
    <select
      ref="input"
      v-model="currentValue"
      :id="name"
      :name="name"
      :required="required"
      :autofocus="autofocus"
      :disabled="disabled"
      class="field__input"
      @focus="handleFocus"
      @blur="handleBlur"
    >
      <option
        v-for="(option, i) in options"
        :key="i"
        :value="option.value"
      >{{ option.label }}</option>
    </select>
    <label
      v-if="label"
      :for="name"
      class="field__label">{{ label }}</label>
    <p
      v-if="error"
      class="field__error">{{ error }}</p>
    <p
      v-else-if="helper"
      class="field__error field__error--helper"
      v-html="helper"/>
    <span class="field__border"/>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'hd-blocks/lang';

export default {
  name: 'HdSelect',
  props: {
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    value: {
      type: [String, Number],
      default: '',
    },
    options: {
      type: Array,
      default: () => [],
    },
    required: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      currentValue: this.value,
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isEmpty() {
      return this.value === null || this.value === undefined || this.value === '';
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--filled': !this.isEmpty,
        'field--valid': false,
        'field--invalid': this.isValid === false,
        'field--disabled': this.disabled,
        'field--hasIcon': this.icon,
      };
    },
  },
  watch: {
    value(value) {
      this.currentValue = value;
    },
    currentValue() {
      this.handleChange();
    },
  },
  methods: {
    handleChange() {
      this.$emit('input', this.currentValue);
      this.validate();
    },
    handleFocus() {
      this.isActive = true;
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && this.isEmpty) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else {
        this.hideError();
      }
      return !this.error;
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'hd-blocks/styles/mixins.scss';

.field {
  &__error {
    width: 100%;
    text-align: left;
    &--helper {
      display: block;
      color: $regent-gray;
    }
  }
  &::after {
    height: 100%;
    top: 0;
    bottom: 0;
  }
}
</style>
