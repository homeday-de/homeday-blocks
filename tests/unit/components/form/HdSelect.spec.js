import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdSelect from '@/components/form/HdSelect.vue';
import ITEMS from '@/stories/mocks/FORM_ITEMS';
import FIELD_CLASSES from './FIELD_CLASSES';

const ERROR_SELECTOR = '.field__error';
const HELPER_SELECTOR = '.field__error--helper';
const ICON_SELECTOR = '.field__icon';

const wrapperBuilder = wrapperFactoryBuilder(HdSelect, {
  props: {
    name: 'testSelect',
    label: 'test select label',
    options: ITEMS,
  },
});

describe('HdSelect', () => {
  it('is rendered as expected', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders with preselected item', () => {
    const ITEM_VALUE = ITEMS[2].value;
    const wrapper = wrapperBuilder({
      props: {
        value: ITEM_VALUE,
      },
    });

    expect(wrapper.find('select').element.value).toBe(ITEM_VALUE);
  });

  it('can change the selected item', () => {
    const ITEM_VALUE = ITEMS[2].value;
    const wrapper = wrapperBuilder({
      props: {
        value: ITEM_VALUE,
      },
    });

    expect(wrapper.find('select').element.value).toBe(ITEM_VALUE);
  });

  it('emits `input` event with the right payload', async () => {
    const ITEM_INDEX = 1;
    const wrapper = wrapperBuilder();

    wrapper.findAll('option').at(ITEM_INDEX).setSelected();

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted('input')[0][0]).toEqual(ITEMS[ITEM_INDEX].value);
  });

  it('allows setting a helper', async () => {
    const wrapper = wrapperBuilder();
    const TEST_HTML = '<b>test</b> test';

    wrapper.vm.showHelper(TEST_HTML);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(HELPER_SELECTOR).element.innerHTML).toEqual(TEST_HTML);
  });

  it('validates requiredness', async () => {
    const wrapper = wrapperBuilder({
      props: {
        value: undefined,
        required: true,
      },
    });

    expect(wrapper.vm.validate()).toBe(false);

    await wrapper.vm.$nextTick();

    expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
    expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

    wrapper.setProps({
      required: false,
    });

    expect(wrapper.vm.validate()).toBe(true);
  });

  it('Supports disabling', () => {
    const wrapper = wrapperBuilder({
      props: {
        disabled: true,
      },
    });

    expect(wrapper.find('select').attributes().disabled).toBe('disabled');
  });

  it('Can render an icon', async () => {
    const wrapper = wrapperBuilder();
    const ICON_PATH = 'fake/icon.svg';

    expect(wrapper.classes()).not.toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);

    wrapper.setProps({
      icon: ICON_PATH,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(true);
  });
});
