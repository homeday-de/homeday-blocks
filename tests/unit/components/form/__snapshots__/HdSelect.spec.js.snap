// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdSelect is rendered as expected 1`] = `
<div class="field field--minimized-label" label="test select label">
  <!---->
  <div class="field__body">
    <div class="field__main"><select id="testSelect" name="testSelect" class="select">
        <option value="jon">
          Jon "Snow"
        </option>
        <option value="daenerys">
          Daenerys
        </option>
        <option value="nightKing">
          The Night King
        </option>
        <option value="arya">
          Arya
        </option>
        <option value="cersei">
          Cersei
        </option>
      </select> <label for="testSelect" class="field__label">
        test select label
      </label>
      <div class="field__input-right">
        <!---->
      </div>
      <div class="field__border"></div>
    </div>
    <p class="field__helper">‚Äç</p>
  </div>
</div>
`;
